packages/convex-helpers/server/zod4.ts

```diff
@@ -0,0 +1,642 @@
+  }
+  if (zodValidator instanceof z.ZodReadonly) {
+    return zodToConvexInternal((zodValidator as any).innerType as z.ZodTypeAny, visited);
+  }
+  if (zodValidator instanceof z.ZodTransform) {
+    const inner = (zodValidator as any).def?.schema as z.ZodTypeAny | undefined;
+    return inner ? zodToConvexInternal(inner, visited) : v.any();
+  }
+  return v.any();
+}
+
+export function zodToConvex<Z extends z.ZodTypeAny | ZodValidator>(zodSchema: Z) {
+  if (zodSchema instanceof z.ZodType) {
+    return zodToConvexInternal(zodSchema);
+  }
+  const out: Record<string, GenericValidator> = {};
+  for (const [k, v_] of Object.entries(zodSchema as Record<string, z.ZodTypeAny>)) {
+    out[k] = zodToConvexInternal(v_);
+  }
+  return out as any;
+}
+
+export function zodToConvexFields<Z extends ZodValidator>(zodShape: Z) {
+  const out: Record<string, GenericValidator> = {};
+  for (const [k, v_] of Object.entries(zodShape)) out[k] = zodToConvexInternal(v_);
+  return out as { [k in keyof Z]: GenericValidator };
+}
+
+// Output mapping (post-transform)
+function zodOutputToConvexInternal(
+  zodValidator: z.ZodTypeAny,
+  visited: Set<z.ZodTypeAny> = new Set(),
+): GenericValidator {
+  if (!zodValidator) return v.any();
+  if (visited.has(zodValidator)) return v.any();
+  visited.add(zodValidator);
+
+  if (zodValidator instanceof z.ZodDefault) {
+    const inner = zodValidator.unwrap() as unknown as z.ZodTypeAny;
+    return zodOutputToConvexInternal(inner, visited);
+  }
+  if (zodValidator instanceof z.ZodTransform) {
+    return v.any();
+  }
+  if (zodValidator instanceof z.ZodReadonly) {
+    return zodOutputToConvexInternal(((zodValidator as any).innerType as unknown) as z.ZodTypeAny, visited);
+  }
+  if (zodValidator instanceof z.ZodOptional) {
+    const inner = zodValidator.unwrap() as unknown as z.ZodTypeAny;
+    return v.optional(asValidator(zodOutputToConvexInternal(inner, visited)));
+  }
+  if (zodValidator instanceof z.ZodNullable) {
+    const inner = zodValidator.unwrap() as unknown as z.ZodTypeAny;
+    return v.union(asValidator(zodOutputToConvexInternal(inner, visited)), v.null());
+  }
+  return zodToConvexInternal(zodValidator, visited);
+}
+
+export function zodOutputToConvex<Z extends z.ZodTypeAny | ZodValidator>(zodSchema: Z) {
+  if (zodSchema instanceof z.ZodType) return zodOutputToConvexInternal(zodSchema);
+  const out: Record<string, GenericValidator> = {};
+  for (const [k, v_] of Object.entries(zodSchema as Record<string, z.ZodTypeAny>)) {
+    out[k] = zodOutputToConvexInternal(v_);
+  }
+  return out as any;
+}
+
+export function zodOutputToConvexFields<Z extends ZodValidator>(zodShape: Z) {
+  const out: Record<string, GenericValidator> = {};
+  for (const [k, v_] of Object.entries(zodShape)) out[k] = zodOutputToConvexInternal(v_);
+  return out as { [k in keyof Z]: GenericValidator };
+}
+
+// Convex -> Zod minimal mapping (for tests)
+export function convexToZod(validator: GenericValidator): z.ZodTypeAny {
```

Comment:
Why is `convexToZod` here marked as being “for tests”? If it serves the same purposes as the equivalent function in v3, it can be useful for non-test purposes (e.g. integrating Convex validators with libraries expecting Zod validators, e.g. the AI SDK)